File: 2025A/P2.m
%% GA优化部分
clear; clc; close all;

% 优化参数配置
nVars = 4;                  
lb = [pi-0.1, 70, 0, 0];         
ub = [pi+0.1, 100, 1, 1];                   

% GA设置
options = optimoptions('ga', ...
    'PopulationSize', 100, ...    
    'MaxGenerations', 15, ...    
    'CrossoverFraction', 0.6, ...
    'MutationFcn', @mutationadaptfeasible, ...  
    'Display', 'iter', ...       
    'PlotFcn', @gaplotbestf);    

[x_opt, fval, exitflag, output] = ga(@ga_fitness, nVars, [], [], [], [], lb, ub, [], intcon, options);
dt_opt = -fval;

fprintf('==================== 遗传算法优化结果 ====================\n');
fprintf('最优参数组合：\n');
fprintf('  角度 th = %.4f 弧度 (%.2f 度)\n', x_opt(1), rad2deg(x_opt(1)));
fprintf('  速度 v = %.4f\n', x_opt(2));
fprintf('  时间 t1 = %.4f\n', x_opt(3));
fprintf('  时间 t2 = %.4f\n', x_opt(4));
fprintf('最优参数对应的dt值：%.4f\n', dt_opt);


%% 原型函数（同supple)
function sol = s2e(a, b, c)
    if a == 0 || (b^2 - 4*a*c) < 0
        sol = [];  
    else
        delta = sqrt(b^2 - 4*a*c);
        x1 = (-b + delta) / (2*a);
        x2 = (-b - delta) / (2*a);
        sol = [x1, x2];  
    end
end


function dt_val = dt(th, v, t1, t2)
    g = 9.8;
    alpha = 3000 / sqrt(101);
    beta = 300 / sqrt(101);
    
    M = [20000 - (t1 + t2)*alpha, 0, 2000 - (t1 + t2)*beta];
    GRD = [17800 - v*(t1 + t2)*cos(th), ...
           v*(t1 + t2)*sin(th), ...
           1800 - 0.5*g*t2^2];
    
    % 生成网格点
    mesh = [];
    for theta = 0:0.1:2*pi - 1e-6
        for r = 0:0.1:7
            x = r * cos(theta);
            y = 200 + r * sin(theta);
            z = 10;
            mesh = [mesh; x, y, z];
        end
    end
    for theta = 0:0.1:2*pi - 1e-6
        for z = 0:0.1:10
            x = 7 * cos(theta);
            y = 200 + 7 * sin(theta);
            mesh = [mesh; x, y, z];
        end
    end
    
    function [t, ifclose] = solvek(t)
        ifclose = false;
        M_new = [M(1) - alpha*t, 0, M(3) - beta*t];
        G_new = [GRD(1), GRD(2), GRD(3) - 3*t];
        
        for i = 1:size(mesh, 1)
            point = mesh(i, :);
            P1 = point(1); P2 = point(2); P3 = point(3);
            M1 = M_new(1); M2 = M_new(2); M3 = M_new(3);
            G1 = G_new(1); G2 = G_new(2); G3 = G_new(3);
            
            a = (M1 - P1)^2 + (M2 - P2)^2 + (M3 - P3)^2;
            b = -2 * ((M1 - P1)*(G1 - P1) + (M2 - P2)*(G2 - P2) + (M3 - P3)*(G3 - P3));
            c = (G1 - P1)^2 + (G2 - P2)^2 + (G3 - P3)^2 - 10^2;
            
            sol = s2e(a, b, c);
            if ~isempty(sol) && any(sol >= 0 - 1e-6) && any(sol <= 1 + 1e-6)
                ifclose = true;
            else
                ifclose = false;
                break;
            end
        end
    end
    
    cntt = 0;
    t_start = 0;
    t_end = 10;
    t_step = 0.002;
    t_list = t_start:t_step:t_end;
    
    for t = t_list
        [~, ifclose] = solvek(t);
        if ifclose
            cntt = cntt + 1;
        end
    end
    
    dt_val = (cntt - 1) * t_step;
end


function fitness = ga_fitness(x)
    th = x(1);   
    v = x(2);    
    t1 = x(3);  
    t2 = x(4);  
    
    dt_val = dt(th, v, t1, t2);
    fitness = -dt_val;  
end

File: 2025A/P2.py
import math
import numpy as np

# 定义原始函数
def s2e(a, b, c):
    if a == 0 or b**2 - 4*a*c < 0:
        return None
    else:
        delta = math.sqrt(b**2 - 4*a*c)
        x1 = (-b + delta) / (2*a)
        x2 = (-b - delta) / (2*a)
        return x1, x2

def dt(th, v, t1, t2):
    # 常数
    g = 9.8
    alpha = 3000 / math.sqrt(101)
    beta = 300 / math.sqrt(101)
    M = (20000 - (t1 + t2) * alpha, 0, 2000 - (t1 + t2) * beta)
    GRD = (17800 - v * (t1 + t2) * math.cos(th), 
           v * (t1 + t2) * math.sin(th), 
           1800 - 0.5 * g * t2 * t2)

    # 画网格
    mesh = []
    for theta in np.arange(0, 2 * math.pi, 0.2*math.pi):
        for r in np.arange(0, 7, 0.5):
            mesh.append((r * math.cos(theta), 200 + r * math.sin(theta), 10))
    for theta in np.arange(0, 2 * math.pi, 0.2*math.pi):
        for z in np.arange(0, 10, 0.5):
            mesh.append((7 * math.cos(theta), 200 + 7 * math.sin(theta), z))

    # 检测函数
    def solvek(t):
        ifclose = False
        M1, M2, M3 = (M[0] - alpha * t, M[1], M[2] - beta * t)
        G1, G2, G3 = (GRD[0], GRD[1], GRD[2] - 3 * t)
        for point in mesh:
            P1, P2, P3 = point
            a = (M1 - P1)**2 + (M2 - P2)** 2 + (M3 - P3)**2
            b = (-2) * ((M1 - P1) * (G1 - P1) + (M2 - P2) * (G2 - P2) + (M3 - P3) * (G3 - P3))
            c = (G1 - P1)** 2 + (G2 - P2)**2 + (G3 - P3)** 2 - 10**2  
            sol = s2e(a, b, c)
            if sol is not None and 0 <= sol[1] <= 1:
                ifclose = True
            else: 
                return (t, False)
        return (t, ifclose)

    # 时间遍历
    cntt = 0
    t_start = 0
    t_end = 10
    t_step = 0.01
    cnttable=[]
    for t in np.arange(t_start, t_end, t_step):
        
        if solvek(t)[1]:
            cntt += 1
            cnttable.append(t)
    
    if cnttable :return (cntt - 1) * t_step,float(cnttable[0]),float(cnttable[-1])
    else:return None

for t1 in np.arange(0, 5, 1):
    for t2 in np.arange(0, 5, 1):
        for theta in np.arange(0,10*np.pi/180, np.pi/180):
            for v in np.arange(70, 140, 10):
                if dt(theta, v, t1, t2) is not None:
                    print(dt(theta, v, t1, t2))